### PR TITLE
Graph Equivalence Changes

### PR DESCRIPTION
**Pull request by emmanvg**
- fix in algorithm to guarantee symmetry
- add different functions for "similarity" and "equivalence"

### CODE REVIEW

File: stix2/environment.py [hunk 1]
```diff
@@ -210,10 +211,10 @@ def semantically_equivalent(obj1, obj2, prop_scores={}, **weight_dict):
                 in the semantic equivalence process
```
Comment (clenk) [inline stix2/environment.py:211]:
```suggestion
                in the similarity process
```

File: stix2/environment.py [hunk 2]
```diff
@@ -229,14 +230,54 @@ def semantically_equivalent(obj1, obj2, prop_scores={}, **weight_dict):
...
+        Args:
+            obj1: A stix2 object instance
+            obj2: A stix2 object instance
+            prop_scores: A dictionary that can hold individual property scores,
+                weights, contributing score, matching score and sum of weights.
+            threshold: A numerical value between 0 and 100 to determine the minimum
+                score to result in successfully calling both objects equivalent. This
+                value can be tuned.
+            weight_dict: A dictionary that can be used to override settings
+                in the semantic equivalence process
```
Comment (clenk) [inline stix2/environment.py:250]:
```suggestion
                in the similarity process
```

File: stix2/equivalence/object/__init__.py [hunk 3]
```diff
@@ -9,30 +11,73 @@
...
     Args:
         obj1: A stix2 object instance
         obj2: A stix2 object instance
         prop_scores: A dictionary that can hold individual property scores,
             weights, contributing score, matching score and sum of weights.
+        threshold: A numerical value between 0 and 100 to determine the minimum
+            score to result in successfully calling both objects equivalent. This
+            value can be tuned.
         weight_dict: A dictionary that can be used to override settings
             in the semantic equivalence process
```
Comment (clenk) [inline stix2/equivalence/object/__init__.py:28]:
```suggestion
            in the similarity process
```

File: stix2/equivalence/object/__init__.py [hunk 4]
```diff
@@ -9,30 +11,73 @@
...
+    """This method returns a measure of similarity depending on how
+    similar the two objects are.
+
+    Args:
+        obj1: A stix2 object instance
+        obj2: A stix2 object instance
+        prop_scores: A dictionary that can hold individual property scores,
+            weights, contributing score, matching score and sum of weights.
+        weight_dict: A dictionary that can be used to override settings
+            in the semantic equivalence process
```
Comment (clenk) [inline stix2/equivalence/object/__init__.py:67]:
```suggestion
            in the similarity process
```

File: stix2/equivalence/object/__init__.py [hunk 5]
```diff
@@ -9,30 +11,73 @@
...
+        see `the Committee Note <link here>`__.
+
+    """
+    similarity_result = object_similarity(obj1, obj2, prop_scores, **weight_dict)
+    if similarity_result >= threshold:
+        return True
+    return False
+
+
+def object_similarity(obj1, obj2, prop_scores={}, **weight_dict):
```
Comment (clenk) [inline stix2/equivalence/object/__init__.py:57]: A fewof the log messages in this function say "semantic equivalence" where "similarity" would be more accurate. Other places we can make the swap are the docstring for `_versioned_checks()` on line 353, the docstring for `reference_check()` on line 376, the docstring for `list_reference_check()` on line 397.

**Review by clenk — COMMENTED** [review]
Looking good :) just a few ideas for small improvements.

File: stix2/environment.py [hunk 6]
```diff
@@ -2,12 +2,13 @@
 import copy
 
 from .datastore import CompositeDataSource, DataStoreMixin
-from .equivalence.graph import graphically_equivalent
+from .equivalence.graph import graph_equivalence, graph_similarity
 from .equivalence.object import (  # noqa: F401
```
[the remark this answers is not in the document]
Comment (emmanvg) [reply stix2/environment.py]: done

File: stix2/equivalence/object/__init__.py (hunk 5 above)
Comment (emmanvg) [reply stix2/equivalence/object/__init__.py:57]: done

File: stix2/equivalence/graph/__init__.py [hunk 7]
```diff
@@ -44,63 +88,63 @@ def graphically_equivalent(ds1, ds2, prop_scores={}, **weight_dict):
...
+    weights["_internal"]["ds1"] = ds1
+    weights["_internal"]["ds2"] = ds2
+
+    for object1, object2 in pairs:
+        iprop_score = {}
+        object1_id = object1["id"]
+        object2_id = object2["id"]
+
+        result = object_similarity(object1, object2, iprop_score, **weights)
+        weights["_internal"]["max_depth"] = depth
```
[the remark this answers is not in the document]
Comment (emmanvg) [reply stix2/equivalence/graph/__init__.py]: done

**Review by clenk — APPROVED** [review]
Looks great, thanks @emmanvg!

**Pull request merged**